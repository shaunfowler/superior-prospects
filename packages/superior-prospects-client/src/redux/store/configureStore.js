import { createStore, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import reduxImmutableStateInvariant from "redux-immutable-state-invariant";
import createHistory from "history/createBrowserHistory";
import { routerMiddleware } from "react-router-redux";
import rootReducer from "../reducers";

export const history = createHistory();

const initialState = {};

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const enhancer = composeEnhancers(
  applyMiddleware(
    reduxImmutableStateInvariant(),
    thunk,
    routerMiddleware(history),
  )
);

const store = createStore(rootReducer, initialState, enhancer);

export default store;
