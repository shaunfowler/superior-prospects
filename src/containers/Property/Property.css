.propertyView {
  padding: 15px;
  flex: 1;
  display: flex;
  flex-direction: column;
}
