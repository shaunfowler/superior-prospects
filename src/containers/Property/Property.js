import React, { Component } from "react";
import {
  Paper,
  Grid,
  Button,
  Typography,
  TextField,
  Select,
  MenuItem,
  Dialog,
  DialogTitle,
  DialogContent,
  DialogContentText,
  DialogActions
} from "material-ui";
import { red } from "material-ui/colors";
import { EditorState } from "draft-js";
import { Editor } from "react-draft-wysiwyg";
import "react-draft-wysiwyg/dist/react-draft-wysiwyg.css";
import { stateToHTML } from "draft-js-export-html";
import { stateFromHTML } from "draft-js-import-html";
import renderHTML from "react-render-html";
import { withRouter } from "react-router";
import {
  trackSaveProperty,
  trackEditButtonClick,
  trackFileUpload,
  trackPropertyDelete,
  trackOpenDeletePropertyModal
} from "../../analytics/propertyAnalytics";
import MediaPanel from "..//MediaPanel";

class Property extends Component {
  constructor(props) {
    super(props);

    this.state = {
      initialized: false,
      errored: false,
      editMode: false,
      showDeletePropertyDialog: false,
      newName: "",
      newDescription: "",
      newLocationRefId: ""
    };
  }

  componentDidMount() {
    const { id } = this.props.match.params;
    const { queryLocations, getProperty, history } = this.props;

    queryLocations();

    getProperty(id)
      .then(response => {
        const { name, body } = this.props.selected;
        document.title = `${name} - Superior Prospects`;
        const contentState = stateFromHTML(body);
        this.setState({
          initialized: true,
          editorState: EditorState.createWithContent(contentState)
        });
      })
      .catch(() => {
        history.push("/");
        this.setState({ initialized: false, errored: true });
      });
  }

  onEditorStateChange = editorState => {
    this.setState({
      editorState
    });
  };

  onNameChange = event => {
    this.setState({ newName: event.target.value });
  };

  onDescriptionChange = event => {
    this.setState({
      newDescription: event.target.value
    });
  };

  onLocationChange = event => {
    this.setState({
      newLocationRefId: event.target.value
    });
  };

  saveProperty = () => {
    const {
      editorState,
      newName,
      newDescription,
      newLocationRefId
    } = this.state;
    const { safeName } = this.props.selected;

    trackSaveProperty(safeName);

    this.props.editProperty(
      Object.assign({}, this.props.selected, {
        body: stateToHTML(editorState.getCurrentContent()),
        name: newName,
        description: newDescription,
        locationRefId: newLocationRefId
      })
    );

    this.setState({
      editMode: false,
      newName: "",
      newDescription: "",
      newLocationRefId: ""
    });
  };

  enterEditMode = () => {
    const { name, safeName, description, locationRefId } = this.props.selected;
    trackEditButtonClick(safeName);
    this.setState({
      editMode: true,
      newName: name,
      newDescription: description,
      newLocationRefId: locationRefId
    });
  };

  onFileDrop = files => {
    const { _id, safeName } = this.props.selected;
    trackFileUpload(safeName);
    const { createMedia } = this.props;
    files.map(file => createMedia(file, _id));
  };

  performDelete = async () => {
    const { selected, deleteProperty, history } = this.props;
    trackPropertyDelete(selected.safeName);
    await deleteProperty(selected._id);
    history.push("/");
    this.hideDeletePropertyDialog();
  };

  showDeletePropertyDialog = () => {
    const { safeName } = this.props.selected;
    trackOpenDeletePropertyModal(safeName);
    this.setState({
      showDeletePropertyDialog: true
    });
  };

  hideDeletePropertyDialog = () => {
    this.setState({
      showDeletePropertyDialog: false
    });
  };

  renderDeleteConfirmationDialog = () => {
    const { selected } = this.props;
    if (!selected) {
      return null;
    }

    return (
      <Dialog
        open={this.state.showDeletePropertyDialog}
        onClose={this.handleClose}
        aria-labelledby="alert-dialog-title"
        aria-describedby="alert-dialog-description"
      >
        <DialogTitle id="alert-dialog-title">
          Delete {selected.name}?
        </DialogTitle>
        <DialogContent>
          <DialogContentText>This cannot be undone.</DialogContentText>
        </DialogContent>
        <DialogActions>
          <Button onClick={this.hideDeletePropertyDialog} color="primary">
            Cancel
          </Button>
          <Button
            onClick={this.performDelete}
            color="default"
            style={{ color: red[400] }}
            autoFocus
          >
            Delete
          </Button>
        </DialogActions>
      </Dialog>
    );
  };

  renderTitle = () => {
    const { selected, isAuthenticated, locations } = this.props;
    const { editMode, newName, newLocationRefId } = this.state;
    const location = locations.find(l => l._id === selected.locationRefId);

    return (
      <div className="paper__title">
        <Typography variant="title">
          {!editMode && selected.name}
          {editMode && (
            <TextField
              className="title"
              onChange={this.onNameChange}
              value={newName}
            />
          )}

          {!editMode && (
            <span className="location">{location && location.name}</span>
          )}
          {editMode &&
            locations && (
              <Select
                className="location-select"
                value={newLocationRefId}
                onChange={this.onLocationChange}
                inputProps={{
                  name: "age",
                  id: "age-simple"
                }}
              >
                {locations.map(l => (
                  <MenuItem key={l._id} value={l._id}>
                    {l.name}
                  </MenuItem>
                ))}
              </Select>
            )}
        </Typography>

        {isAuthenticated &&
          !editMode && (
            <Button
              color="default"
              style={{ color: red[400] }}
              onClick={() => this.showDeletePropertyDialog()}
            >
              Delete
            </Button>
          )}

        {editMode && (
          <Button color="primary" onClick={() => this.saveProperty()}>
            Save
          </Button>
        )}
        {isAuthenticated &&
          !editMode && (
            <Button color="primary" onClick={() => this.enterEditMode()}>
              Edit
            </Button>
          )}
      </div>
    );
  };

  renderDescription = () => {
    const { editMode, newDescription } = this.state;
    const { selected } = this.props;
    return (
      <div className="description">
        {!editMode && <Typography>{selected.description}</Typography>}
        {editMode && (
          <TextField
            type="text"
            label="Description"
            multiline={true}
            rowsMax={5}
            className="subtitle"
            value={newDescription}
            onChange={this.onDescriptionChange}
          />
        )}
      </div>
    );
  };

  renderBody = () => {
    const { editorState, editMode } = this.state;
    const { selected } = this.props;
    return (
      <div className="body">
        {editMode ? (
          <Editor
            editorState={editorState}
            toolbarClassName="toolbarClassName"
            wrapperClassName="wrapperClassName"
            editorClassName="editorClassName"
            onEditorStateChange={this.onEditorStateChange}
          />
        ) : (
          selected.body && renderHTML(selected.body)
        )}
      </div>
    );
  };

  renderMedia = () => {
    const { selected, isAuthenticated } = this.props;
    return (
      <MediaPanel
        media={selected.media}
        onFileDrop={this.onFileDrop}
        isAuthenticated={isAuthenticated}
      />
    );
  };

  render() {
    return (
      <div className="propertyView container">
        <Paper className="paper" elevation={1}>
          {this.state.initialized && (
            <Grid container spacing={24}>
              <Grid item xs={12} style={{ paddingTop: 0, paddingBottom: 0 }}>
                {this.renderTitle()}
                {this.renderDescription()}
              </Grid>
              <Grid item xs={12} sm={12} md={8} lg={8} xl={9}>
                {this.renderBody()}
              </Grid>
              <Grid item xs={12} sm={12} md={4} lg={4} xl={3}>
                {this.renderMedia()}
              </Grid>
            </Grid>
          )}
        </Paper>
        {this.renderDeleteConfirmationDialog()}
      </div>
    );
  }
}

export default withRouter(Property);
