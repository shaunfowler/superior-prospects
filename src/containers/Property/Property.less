.propertyView {
  display: flex;
  margin-top: (-64px - 54px);

  .description {
    padding: 15px 15px 0 15px;
  }

  .body {
    padding: 0 15px;
  }

  .media {
    padding: 15px;
  }

  .body {
    font-size: 14px;
    line-height: 20px;
    color: #333;
  }

  textarea {
    width: 100%;
  }

  .title {
    width: 50%;
  }

  .location {
    margin-left: 10px;
    font-size: 16px;
    color: #999;
  }

  .location-select {
    position: relative;
    bottom: -3px;
    margin-left: 20px;
  }

  .subtitle {
    width: 100%;
  }

  a {
    color: #3273dc;
  }
}
