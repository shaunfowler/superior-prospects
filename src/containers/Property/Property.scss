.propertyView {
  flex: 1;
  padding: 20px 0;
}
