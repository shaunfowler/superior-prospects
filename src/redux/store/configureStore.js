import { createStore, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import reduxImmutableStateInvariant from "redux-immutable-state-invariant";
import createHistory from "history/createBrowserHistory";
import { routerMiddleware } from "react-router-redux";
import rootReducer from "../reducers";
import routeTrackerMiddleware from "../middleware/routeTrackerMiddleware";

export const history = createHistory();

const initialState = {};

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const enhancer = composeEnhancers(
  applyMiddleware(
    reduxImmutableStateInvariant(),
    thunk,
    routerMiddleware(history),
    routeTrackerMiddleware
  )
);

const store = createStore(rootReducer, initialState, enhancer);

export default store;
