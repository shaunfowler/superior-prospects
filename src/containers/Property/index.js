import { connect } from "react-redux";
import Property from "./Property";
import { getProperties } from "../../redux/actions/propertiesActions";
import "./Property.css";

const mapStateToProps = state => {
  return {
    properties: state.properties
  };
};

const mapDispatchToProps = dispatch => {
  return {
    getProperties: () => dispatch(getProperties())
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Property);
