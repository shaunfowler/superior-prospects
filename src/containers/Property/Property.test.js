import React from 'react';
import ReactDOM from 'react-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Property from './Property';

const mockStore = configureStore([thunk]);

it('renders without crashing', () => {
  const initialState = {};
  const initialProps = {
    match: { params: {} },
    properties: { list: [{ id: 1234 }] },
    getProperties: () => Promise.resolve([])
  };
  const store = mockStore(initialState);
  const div = document.createElement('div');
  ReactDOM.render(<Property store={store} {...initialProps} />, div);
});
