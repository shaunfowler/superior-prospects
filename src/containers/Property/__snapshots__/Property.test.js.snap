// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Property container renders without crashing 1`] = `
<Route
  render={[Function]}
/>
`;
